File: src/adapters/embedded/projection.ts
/**
 * SqliteProjection — Phase 2 Embedded Access. An FTS5 index over a repository's
 * Canonical state. The projection is REBUILDABLE and NON-AUTHORITATIVE:
 *  - it only LOCATES object ids (via FTS5); values are always read back from
 *    the canonical repository (read(objectId, commit)).
 *  - it records its basis (repository + commit) and reports lag behind head.
 *  - deleting/rebuilding the index never touches canonical state (K-19).
 */

import { createRequire } from "node:module";
import type { CommitIdentity, KnowledgeValue, ObjectIdentity, RepositoryIdentity } from "../../contracts/index.ts";

// node:sqlite is a very recent Node built-in that Vite's resolver does not yet
// recognize; load it through createRequire to bypass static analysis.
interface SqliteStatement {
  run(...params: unknown[]): unknown;
  get(...params: unknown[]): unknown;
  all(...params: unknown[]): unknown[];
}
interface SqliteDatabase {
  exec(sql: string): void;
  prepare(sql: string): SqliteStatement;
}

const require = createRequire(import.meta.url);
const { DatabaseSync } = require("node:sqlite") as {
  DatabaseSync: new (location: string) => SqliteDatabase;
};

/** Minimal read surface both MemoryRepository and FileGitRepository satisfy. */
export interface ReadableRepository {
  readonly repositoryId: RepositoryIdentity;
  head(): CommitIdentity;
  list(commitId: CommitIdentity): KnowledgeValue[];
  read(objectId: ObjectIdentity, commitId: CommitIdentity): KnowledgeValue;
}

export interface IndexDescriptor {
  readonly basisRepository: RepositoryIdentity;
  readonly basisCommit: CommitIdentity;
  readonly objectCount: number;
  readonly headCommit: CommitIdentity;
  readonly lagBehindHead: boolean;
}

export class SqliteProjection {
  private db: SqliteDatabase;
  private basisRepository: RepositoryIdentity = "";
  private basisCommit: CommitIdentity = "";

  constructor() {
    this.db = new DatabaseSync(":memory:");
  }

  /** (Re)build the FTS5 index from a pinned commit. */
  build(repository: ReadableRepository, commit: CommitIdentity): void {
    this.db.exec("DROP TABLE IF EXISTS idx");
    this.db.exec("CREATE VIRTUAL TABLE idx USING fts5(object_id UNINDEXED, value_text)");
    const insert = this.db.prepare("INSERT INTO idx(object_id, value_text) VALUES (?, ?)");
    for (const value of repository.list(commit)) {
      insert.run(value.address.objectId, JSON.stringify(value.value));
    }
    this.basisRepository = repository.repositoryId;
    this.basisCommit = commit;
  }

  /** FTS5 word-level search; values are read back from Canonical (not the index). */
  search(repository: ReadableRepository, query: string): KnowledgeValue[] {
    const fts = query
      .trim()
      .split(/\s+/)
      .filter(Boolean)
      .map((w) => `"${w.replaceAll('"', '""')}"`)
      .join(" AND ");
    const rows = this.db.prepare("SELECT object_id FROM idx WHERE value_text MATCH ?").all(fts) as {
      object_id: string;
    }[];
    return rows.map((r) => repository.read(r.object_id, this.basisCommit));
  }

  describeIndex(repository: ReadableRepository): IndexDescriptor {
    const row = this.db.prepare("SELECT count(*) AS c FROM idx").get() as { c: number };
    const head = repository.head();
    return {
      basisRepository: this.basisRepository,
      basisCommit: this.basisCommit,
      objectCount: row.c,
      headCommit: head,
      lagBehindHead: head !== this.basisCommit,
    };
  }
}

File: reader/projection.go
package reader

import (
	"encoding/json"
	"strings"

	"kc/kernel"
	"kc/repository"
)

type IndexDescriptor struct {
	BasisRepository kernel.RepositoryID `json:"basisRepository"`
	BasisCommit     kernel.CommitID     `json:"basisCommit"`
	ObjectCount     int                 `json:"objectCount"`
	HeadCommit      kernel.CommitID     `json:"headCommit"`
	LagBehindHead   bool                `json:"lagBehindHead"`
	AccessDigest    kernel.Digest       `json:"accessDigest,omitempty"`
	Mode            string              `json:"mode,omitempty"`
	Cause           string              `json:"cause,omitempty"`
}

type indexRow struct {
	objectID  kernel.ObjectID
	valueText string
}

// Projection is an Access Projection (design 7.3 / K-19): rebuildable, not
// Canonical, not a KnowledgeRef source. This FileGit profile is in-memory
// AND-contains (lexical). Hits hydrate from the pinned Canonical commit.
// Not EvaluationProjection (Refine field whitelist) and not AspectSelector.
type Projection struct {
	rows            []indexRow
	basisRepository kernel.RepositoryID
	basisCommit     kernel.CommitID
	selector        *repository.AspectSelector
}

func NewProjection() *Projection {
	return &Projection{}
}

func (p *Projection) Build(repo repository.Repository, commit kernel.CommitID, selector *repository.AspectSelector) error {
	p.selector = selector
	listed, err := repo.List(commit)
	if err != nil {
		return err
	}
	p.rows = p.rows[:0]
	for _, value := range listed {
		text := repository.SelectAspects(value.Value, value.Units, selector)
		b, _ := json.Marshal(text)
		p.rows = append(p.rows, indexRow{objectID: value.Address.ObjectID, valueText: string(b)})
	}
	p.basisRepository = repo.ID()
	p.basisCommit = commit
	return nil
}

func (p *Projection) Search(repo repository.Repository, query string) ([]repository.KnowledgeValue, error) {
	words := tokenize(query)
	var hits []repository.KnowledgeValue
	for _, row := range p.rows {
		if !allWords(row.valueText, words) {
			continue
		}
		value, err := repo.Read(row.objectID, p.basisCommit)
		if err != nil {
			return nil, err
		}
		if p.selector != nil {
			value.Value = repository.SelectAspects(value.Value, value.Units, p.selector)
		}
		hits = append(hits, value)
	}
	return hits, nil
}

func (p *Projection) DescribeIndex(repo repository.Repository) (IndexDescriptor, error) {
	head, err := repo.Head("")
	if err != nil {
		return IndexDescriptor{}, err
	}
	return IndexDescriptor{
		BasisRepository: p.basisRepository,
		BasisCommit:     p.basisCommit,
		ObjectCount:     len(p.rows),
		HeadCommit:      head,
		LagBehindHead:   head != p.basisCommit,
	}, nil
}

func tokenize(query string) []string {
	var words []string
	for _, w := range strings.Fields(strings.TrimSpace(query)) {
		if w != "" {
			words = append(words, strings.ToLower(w))
		}
	}
	return words
}

func allWords(text string, words []string) bool {
	lower := strings.ToLower(text)
	for _, w := range words {
		if !strings.Contains(lower, w) {
			return false
		}
	}
	return len(words) > 0
}
